compiler/compiler.go

```diff
@@ -337,8 +337,20 @@ func (c *Compiler) Compile(mainPath string) []error {
 
 	c.loadASTComments(lprogram)
 
-	// Declare all functions.
+	// Declare functions with a body.
+	// This is necessary so that the correct types are used across linkname.
 	for _, f := range c.ir.Functions {
+		if f.Blocks == nil {
+			continue
+		}
+		frames = append(frames, c.parseFuncDecl(f))
+	}
+
+	// Declare all functions that do not have a body.
+	for _, f := range c.ir.Functions {
+		if f.Blocks != nil {
+			continue
+		}
```

Comment:
This does not appear to be necessary: the tests still pass without this change.

Reply:
I have confirmed that this is necessary. At the moment I don't really see a reasonable way to test this.

Basically, the interrupt synchronization primitive package on `intsched` demonstrates this behavior, as the `machine/sync` package gets initialized before `runtime`, which causes the compiler to crash later due to assertions. This is necessary to ensure that the functions are initialized in the correct order.